### ISSUE TITLE
Dynamic audio sinks

### ISSUE THREAD

**Issue by liamw9534** [mopidy/mopidy#1118]
I apologize in advance if this is an existing feature request or something that has been asked before but I thought I'd share this idea for a feature enhancement that I was working on during last year...

The basic principle is to extend the Mopidy ``Core`` and ``Audio`` classes to allow for the gst sink to be dynamically configured and also for the possibility of multiple sinks to be added.  No assumption here is made about whether multiple sinks need to be synchronized or not (I note this topic is discussed elsewhere).

This feature could co-exist with the existing audio sink behaviour by introducing a new boolean property called as ``dynamic_sink_enable`` with a default value of ``false``.  When set to ``true`` the existing ``audio_sink`` property would be ignored and audio playback would utilize dynamic audio sinks.

A dynamic audio sink would be realized through the use of an extension which provides functionality enabling a specific type of audio output subsystem (or technology) to be detected, enumerated, connected or disconnected (e.g., ALSA, Pulse, Bluetooth, WiFi, other networked audio output).  It would be the role of the extension to allow sinks to be reported for end user selection.  This would be done through HTTP API and relies on a separate feature that I have been working on i.e., the ``service`` framework which allows extensions to publish their APIs over HTTP [1].

The Core/Audio API would be required to support two new functions ``add_audio_sink()`` and ``remove_audio_sink()``.  A ``sink`` in this context is a Python object that is sub-classed from ``GstBin`` and contains a ``GstSink`` as its final element.  It must also supports the required audio input formats and functionality to render audio to some output device.  In practice, each sink object would most likely be a wrapper to an existing gst plugin component e.g., alsa, pulse, bluetooth, udp socket, tcp socket, etc.  The wrapper would have certain requirements, such as prepending a ``queue`` element to allow fitting into a ``tee``.  These requirement could, in principle, be made hidden from the extension developer through suitable subclassing.

The audio sink subsystem would be augmented with a new ``AudioOutput`` class [2] which would encapsulate the new dynamic behaviours.  When the dynamic sink property is enabled, ``Audio`` would create an instance of ``AudioOutput`` and set this as its default audio sink as part of the gst playbin pipeline creation.  In practice, ``AudioOutput`` is merely a container class which, by default, creates a ``tee`` with a single ``queue ! fakesink`` arm.  This step is necessary to have a fully "closed" pipeline on start-up.

The container is then updated by using ``add_audio_sink`` and ``remove_audio_sink`` methods which allows new sinks to be added and removed.  This involves ``AudioOutput`` dynamically adding/removing elements from its local ``tee``.

Individual audio sink instances must be uniquely enumerated and identified by the extension class for the API to work properly.  As an example, an ALSA audio extension might enumerate its sinks thus: ``alsa:dev:0``, ``alsa:dev:1``, etc.  By prepending each audio sink with ``alsa`` there is no risk of conflict with other audio sink extensions.  Moreover, the sink name acts as an identifier that can be used both by the UI and the ``AudioOutput`` API.

I do have an outline implementation of the above working but it is prior to mopidy v1.0.  The trickiest part was getting dynamic insertion/removal from ``tee`` working reliably without causing the pipeline to lock up.  The implementation works for GStreamer 0.10 (tested with Alsa [3] and Pulse [4] sinks) but GStreamer 1.0 is a different beast and therefore I wouldn't expect it to work there.

As a minimum, I hope this can provoke some discussion around the concept and API proposal even if the implementation is not deemed suitable.

References:

[1] https://github.com/liamw9534/mopidy/blob/liamw9534_service_framework/mopidy/service/__init__.py
[2] https://github.com/liamw9534/mopidy/blob/liamw9534_service_framework/mopidy/audio/output.py
[3] https://github.com/liamw9534/mopidy-alsa
[4] https://github.com/liamw9534/mopidy-pulseaudio

**Comment by adamcik** [comment]
I don't have time to address this in detail right now, but this is essentially #790 :-) There is some ground work with having a tee etc, but getting it really really reliable with GStreamer 0.10 is _hard_ and there are a lot of fixes in GStreamer 1.0 that resolves a lot of the issues.

Back home I have some tests / PoC music players that implement these things in isolation from the rest of Mopidy make it easier to debug and test. I'll put up a gist with it/them when I get home.

I'm very happy you are looking at this, as getting this right requires quite a bit of GStreamer foo and having two of us that understand it somewhat and bounce things back and forth should help make this reality :-)

**Comment by liamw9534** [comment]
Yes, I had a vague recollection that you had done something on this before!  I just couldn't find the right issue search keywords ;-)

Anyway, what you are saying is pretty much exactly what I had to do.  With Gstreamer 0.10, I did find that with some sinks it was not reliable (e.g., ``a2dpsink``) and would randomly fail when the pipeline is running.  But with an ``alsasink`` I could never break it.  So, I suspect there is a strong dependency with how the output sink is written as to whether it works at all in 0.10 but couldn't find a specific design-rule when reviewing the C source code of various audio sinks.

I'm not sure if there is a need any longer to worry about gstreamer 0.10?  To be honest, I haven't yet re-established my dev env for mopidy 1.0 nor looked at gstreamer 1.0 in great detail so that'll be something I have to do soon.

Maybe we could agree first upon the general framework and API?

**Comment by kingosticks** [comment]
Just to clarify, Mopidy 1.0 still uses gstreamer 0.10.

**Comment by liamw9534** [comment]
Ahhh, ok - got it.

**Comment by adamcik** [comment]
https://gist.github.com/adamcik/de74b67847b8098238f5 (not that I've checked what state I left this in)

**Comment by adamcik** [comment]
As for the API, on a high level core should be able to list outputs, toggle them add new/remove them as they appear. Outputs need to have a name so they can map to MPD outputs. The simplest output will likely just be "generic" GStreamer output that lets you set the equivalent of today's `audio/output` setting. More advanced ones would be HTTP streaming or other specialized ones that have some "preset" GStreamer elements, or need more configuration.

So far I'm completely ignoring the fact that some outputs will need to be configured. In all likelihood, we can progress in the following stages so we don't have to get stuck with a monster PR that is harder to get in:

1. Multiple outputs that can only be configured at startup
2. Multiple outputs that can be toggeled on or off at runtime
3. Dynamic outputs that can be added runtime
4. Runtime configuration of said dynamic outputs

**Comment by liamw9534** [comment]
I like the idea of creating a standalone test harness.  It gives a lot more control over testing and debug.

It looks like we arrived at  similar solutions - I will do a more thorough comparison tomorrow and also run some tests to check behaviour.

**Comment by adamcik** [comment]
See #790 for a lot of links and other reference material, as I've no closed that issue as we don't need two for the same thing.

**Comment by foo-git** [comment]
I'm also interested in having this feature and read about it in #121/#790. Mopidy is now using gstreamer1.0, apparently removing some "sharp edges". What are the current known issues which will make the implementation difficult?

**Comment by kingosticks** [comment]
I did a little bit of work trying to add a (static, not dynamic) streaming HTTP output but I couldn't get it working in a few scenarios (including appsrc). I didn't debug enough to find the issue and I've seen been otherwise occupied. My point being that there are possibly still some issues with GST 1.0 but I don't think we know what they are. It'd be great for someone interested to take this issue on in way [suggested](https://github.com/mopidy/mopidy/issues/1118#issuecomment-90986376).

Note you can already achieve multiple static sinks using a GST tee element but I'm not convinced all sinks support that to the same standard and there are differences between v1.14 and v1.16 (probably v1.18 also). 

Out of interest, what's your use-case?

**Comment by foo-git** [comment]
Background for use case: I have in three rooms each a set of speakers connected via audio cable to a central server running mopidy. Each room is modeled on the server as an alsasink.

Use case is to manually toggle off the audio stream to a certain room/sink while keeping the audio stream to another room/sink active. For example, a user wishes to do something in a room without music, but another user in another room still wants to listen to some music.

Basically it's the same behaviour mpd is providing right now. In the [comment you mentioned](https://github.com/mopidy/mopidy/issues/1118#issuecomment-90986376) my use case requires only step 1+2.

**Comment by kingosticks** [comment]
OK, that's a nice simple case since it's all the same sink type and alsasink is one of the better ones. Unfortunately we'll probably have to document it very clearly that mixing other sink types is going to have mixed results, and maybe there will be fixes required for mitigating that but that shouldn't stop us. However, this will need to work with appsrc.

**Comment by adamcik** [comment]
I think this kept failing with Spotify due some buffer issues in the gstreamer pipeline. If memory serves this was fixed at some point before libspotify got killed. I'm not sure what the state is now that we are switching to a Spotify element.

But it might make sense to revisit this between the Spotify changes and never gstreamer.

**Comment by adamcik** [comment]
@janiversen if you are interested in multiroom see this issue and #790 for the prior state (in case you haven't seen them yet). I think this is doable now, per my last comment about the Spotify bug being fixed. I was close to having it working way back when, at least to the point of having multiple outputs.

Then to have good multiroom there is of course sync issues and other tricky things to get right...

**Comment by janiversen** [comment]
@adamcik thanks for the pointer, I had not seen this one. It sounds like the way to go.

Synchronization is a very big issue if it needs to be perfect, some vendors have a loop back scheme to calculate the delay, but first I need to do the basic.
